components/director/internal/tenantfetchersvc/provisioner.go

```diff
@@ -12,121 +11,112 @@ import (
 	"github.com/pkg/errors"
 )
 
-const regionLabelKey = "region"
+const autogeneratedTenantProvider = "autogenerated"
 
 //go:generate mockery --name=TenantService --output=automock --outpkg=automock --case=underscore --unroll-variadic=False
 type TenantService interface {
 	GetInternalTenant(ctx context.Context, externalTenant string) (string, error)
 	CreateManyIfNotExists(ctx context.Context, tenantInputs ...model.BusinessTenantMappingInput) error
-	SetLabel(ctx context.Context, labelInput *model.LabelInput) error
+}
+
+type TenantProvisioningRequest struct {
+	AccountTenantID    string
+	SubaccountTenantID string
+	CustomerTenantID   string
+	Subdomain          string
+	Region             string
 }
 
 type provisioner struct {
-	tenantSvc TenantService
+	tenantSvc      TenantService
+	tenantProvider string
 }
 
-func NewTenantProvisioner(tenantSvc TenantService) *provisioner {
+func NewTenantProvisioner(tenantSvc TenantService, tenantProvider string) *provisioner {
 	return &provisioner{
-		tenantSvc: tenantSvc,
+		tenantSvc:      tenantSvc,
+		tenantProvider: tenantProvider,
 	}
 }
 
-func (p *provisioner) ProvisionTenant(ctx context.Context, tenant model.BusinessTenantMappingInput) error {
-	if tenant.Type == tenantEntity.TypeToStr(tenantEntity.Subaccount) {
-		return fmt.Errorf("tenant with ID %s is of type %s and supports only regional provisioning", tenant.ExternalTenant, tenant.Type)
+func (p *provisioner) ProvisionTenants(ctx context.Context, request TenantProvisioningRequest) error {
+	if len(request.SubaccountTenantID) > 0 {
+		return fmt.Errorf("tenant with ID %s is of type %s and supports only regional provisioning", request.SubaccountTenantID, tenantEntity.Subaccount)
 	}
 
-	return p.createTenant(ctx, tenant)
-}
-
-func (p *provisioner) ProvisionRegionalTenant(ctx context.Context, tenant model.BusinessTenantMappingInput, region string) error {
-	if err := p.createTenant(ctx, tenant); err != nil {
+	tenants, err := p.tenantsFromRequest(ctx, request)
+	if err != nil {
 		return err
 	}
 
-	intTenantID, err := p.tenantSvc.GetInternalTenant(ctx, tenant.ExternalTenant)
+	return p.tenantSvc.CreateManyIfNotExists(ctx, tenants...)
+}
+
+func (p *provisioner) ProvisionRegionalTenants(ctx context.Context, request TenantProvisioningRequest) error {
+	tenants, err := p.tenantsFromRequest(ctx, request)
 	if err != nil {
 		return err
 	}
+	return p.tenantSvc.CreateManyIfNotExists(ctx, tenants...)
+}
 
-	label := &model.LabelInput{
-		Key:        regionLabelKey,
-		Value:      region,
-		ObjectID:   intTenantID,
-		ObjectType: model.TenantLabelableObject,
-	}
-	if err := p.tenantSvc.SetLabel(ctx, label); err != nil {
-		return errors.Wrapf(err, "while setting region label for tenant with external ID %s", tenant.ExternalTenant)
-	}
+func (p *provisioner) tenantsFromRequest(ctx context.Context, request TenantProvisioningRequest) ([]model.BusinessTenantMappingInput, error) {
```

Comment:
Suggestion:

This way it is a bit easier to follow for me, however, it makes more DB calls. I am not sure which is better, so I am fine with both approaches.

```go
func (p *provisioner) tenantsFromRequest(ctx context.Context, request TenantProvisioningRequest) ([]model.BusinessTenantMappingInput, error) {
	tenants := make([]model.BusinessTenantMappingInput, 0, 3)
	customerID := request.CustomerTenantID
	accountID := request.AccountTenantID

	if len(request.CustomerTenantID) > 0 {
		customerInternalID, err := p.tenantSvc.GetInternalTenant(ctx, request.CustomerTenantID)
		if err != nil && !apperrors.IsNotFoundError(err) {
			return nil, errors.Wrapf(err, "while checking if parent of type customer already exists")
		}
		if len(customerInternalID) == 0 {
			tenants = append(tenants, p.newCustomerTenant(request.CustomerTenantID))
		} else {
			customerID = customerInternalID
		}
	}

	if len(request.AccountTenantID) > 0 {
		accountInternalID, err := p.tenantSvc.GetInternalTenant(ctx, request.AccountTenantID)
		if err != nil && !apperrors.IsNotFoundError(err) {
			return nil, errors.Wrapf(err, "while checking if parent of type customer already exists")
		}
		if len(accountInternalID) == 0 {
			accToInsert := p.newAccountTenant(request.AccountTenantID, customerID, request.Subdomain, request.Region)
			if len(request.SubaccountTenantID) > 0 { // This means that the request is for Subaccount provisioning, therefore the subdomain and the region are for the subaccount and not for the GA
				accToInsert.Subdomain = ""
				accToInsert.Region = ""
			}
			tenants = append(tenants, accToInsert)
		} else {
			accountID = accountInternalID
		}
	}

	if len(request.SubaccountTenantID) > 0 {
		tenants = append(tenants, p.newSubaccountTenant(request.SubaccountTenantID, accountID, request.Subdomain, request.Region))
	}
	return tenants, nil
}
```